src/query/src/part_sort.rs

```diff
@@ -0,0 +1,544 @@
+            first
+        } else {
+            internal_err!("No children found")?
+        };
+        Ok(Arc::new(Self::try_new(
+            self.expression.clone(),
+            self.fetch,
+            self.ranges.clone(),
+            new_input.clone(),
+        )?))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> datafusion_common::Result<DfSendableRecordBatchStream> {
+        self.to_stream(context, partition)
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metrics.clone_inner())
+    }
+}
+
+struct PartSortStream {
+    /// Memory pool for this stream
+    reservation: MemoryReservation,
+    buffer: Vec<DfRecordBatch>,
+    expression: PhysicalSortExpr,
+    fetch: Option<usize>,
+    produced: usize,
+    input: DfSendableRecordBatchStream,
+    input_complete: bool,
+    schema: SchemaRef,
+    /// The current PartitionRange's index
+    range_idx: usize,
+}
+
+impl PartSortStream {
+    fn new(
+        context: Arc<TaskContext>,
+        sort: &PartSortExec,
+        input: DfSendableRecordBatchStream,
+    ) -> Self {
+        Self {
+            reservation: MemoryConsumer::new("PartSortStream".to_string())
+                .register(&context.runtime_env().memory_pool),
+            buffer: Vec::new(),
+            expression: sort.expression.clone(),
+            fetch: sort.fetch,
+            produced: 0,
+            input,
+            input_complete: false,
+            schema: sort.input.schema(),
+            range_idx: 0,
+        }
+    }
+}
+
+impl PartSortStream {
+    fn remaining_fetch(&self) -> Option<usize> {
+        self.fetch.map(|fetch| fetch - self.produced)
+    }
+
+    /// Sort and clear the buffer and return the sorted record batch
+    ///
+    /// this function should return None if RecordBatch is empty
+    fn sort_buffer(&mut self) -> datafusion_common::Result<Option<DfRecordBatch>> {
+        if self.buffer.is_empty() || self.buffer.iter().map(|r| r.num_rows()).sum::<usize>() == 0 {
+            return Ok(None);
+        }
+        let mut sort_columns = Vec::with_capacity(self.buffer.len());
+        let mut opt = None;
+        for batch in self.buffer.iter() {
+            let sort_column = self.expression.evaluate_to_sort_column(batch)?;
+            opt = opt.or(sort_column.options);
+            sort_columns.push(sort_column.values);
+        }
+
+        let sort_column =
+            concat(&sort_columns.iter().map(|a| a.as_ref()).collect_vec()).map_err(|e| {
+                DataFusionError::ArrowError(
+                    e,
+                    Some(format!("Fail to concat sort columns at {}", location!())),
+                )
+            })?;
+
+        let limit = self.remaining_fetch().and_then(|f| {
+            let len = sort_column.len();
+            if f < len {
+                Some(f)
+            } else {
+                None
+            }
+        });
```

Comment:
> Will it be more helpful if we apply the limit to each partition range in this stream instead of all partition ranges?

I don't quite understand, why is that more useful?

Reply:
The limit across all partition ranges is meaningless. Rows from different partition ranges are unsorted and may be overlapping.

Reply:
> The limit across all partition ranges is meaningless. Rows from different partition ranges are unsorted and may be overlapping.

Due to this PartSort is mostly input for WindowSort, having a fetch can reduce the numbers of rows WindowSort have to merge sort if fetch truncate a large RecordBatch in the middle(The latter part of that RecordBatch wouldn't need be send to WindowSort), this might just be pre mature optimization though

Reply:
> I wonder if we can apply the fetch to each partition range which can reduce the rows to sort in the window sort. But maybe this isn't a necessary optimization. It's fine to remove the fetch here.

I think that optimization will be even more complex than WindowSort since you need to keep track of all Part's order and index, essentially doing a merge sort in here to found out exactly how many slice of recordbatch to send, probably not worth the trouble unless we properly benchmark how often did WindowSort encounter this and how much does it cost